productions/production.py

```diff
@@ -25,10 +26,21 @@ def _right_side(self) -> HyperGraph:
         pass
 
     @abstractmethod
-    def apply(self, graph: HyperGraph) -> (bool, HyperGraph):
-        """Abstract method to apply the production."""
+    def transform(self, graph: HyperGraph, node_map: dict[Node, Node], left_side: HyperGraph) -> (bool, HyperGraph):
+        """Abstract method to transform the main graph."""
         pass
 
+    def apply(self, graph: HyperGraph) -> (bool, HyperGraph):
+        left_side = self.get_left_side()
+        node_map = self.check(graph, left_side)
+
+        if node_map:
+            reversed_node_map = {v: k for k, v in node_map.items() if isinstance(v, Node) and isinstance(k, Node)}
+
+            return self.transform(graph, reversed_node_map, left_side)
```

Comment:
this line is to make further checks possible, not useful now, but may save from edge cases in the future